src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,407 @@
+answer, so if they are ordering a single cake and it might be ready soon, then
+great, let that thread focus on replying about the one cake, and when that cake
+is doing the connection is closed. If the client wants multiple status updates,
+or is interested in multiple cakes, this might not be the way to go.
+
+Personal experiences tell me this sort of blocking activity can lead to quite a
+mess when you have more than a smattering of users. Blocking an entire web
+thread for 3 seconds can certainly be a problem when you have a lot of clients
+asking about a lot of cakes.
+
+For example, in Rails land this is a pretty common server config:
+
+[source,ruby]
+----
+workers Integer(ENV['WEB_CONCURRENCY'] || 2)
+threads_count = Integer(ENV['RAILS_MAX_THREADS'] || 5)
+threads threads_count, threads_count
+
+rackup      DefaultRackup
+port        ENV['PORT']     || 3000
+environment ENV['RACK_ENV'] || 'development'
+----
+
+If there was only one dyno running (a.k.a instance, pod, whatever) then 2 * 5 =
+10 and thats only 10 cake orders happening at the same time and you've run out of
+threads for other things, like accepting payments, or registering a user.
+
+Even having a slightly bulkier setup, with more dynos, and more processes, and more threads,
+its simply a numbers game where you are throwing more money at the situation.
+
+=== Web Hooks
+
+Both these polling approaches are essentially a pull model, so lets look at a
+push model: Web Hooks. Clients register a "callback URL" through a
+GUI or API, and they may also specify specific types of information they are
+interested in receiving. When relevant events happen on the server, it will fire
+a payload at that URL, which usually takes the form of a HTTP
+POST request.
+
+.Examples of Web Hooks in popular APIs.
+- https://api.slack.com/incoming-webhooks[Slack API > Incoming Webhooks]
+- https://stripe.com/docs/webhooks[Stripe API > Webhooks]
+
+A few years ago my friend and I made a website that sold silly programmer joke
+t-shirts. We created the whole thing over beers, and we made some whopping
+mistakes. Our payment acceptance sequence looked like the synchronous one above, as
+we didn't care about speed at that early stage. The biggest mistake was that
+when the payment API gave us a positive response to the `POST` request, we
+immediately considered the payment good and marked the shirts for dispatch.
+
+It turns out there was a web hook we were meant to listen out for, and this
+would confirm if the payment had been accepted, or not... Seeing as we totally
+ignored that, there were a few t-shirts that went out for free! Oops.
+
+This was simple enough, we added a `/callback` endpoint, registered it with the
+payment gateway, and they would fire updates at it a few seconds or minutes
+after a payment attempt was made. We simply marked orders as pending in the mean
+time, made a order status page, and they would get an email later. We could have
```

Comment:
```suggestion
time, made an order status page, and they would get an email later. We could have
```

Reply:
I think this could be made a bit clearer, something like:
```
This was simple enough to fix, we added a `/callback` endpoint, registered it with the
payment gateway, and they would fire updates at it once the payment was confirmed. We simply marked orders as pending in the meantime and only marked them for dispatch (and sent emails to the customers) once we received the confirmation via web hooks. We could have...
```